pkg/cmd/server/start/start_node.go

```diff
@@ -281,13 +320,295 @@ func (o NodeOptions) CreateNodeConfig() error {
 	if err := createNodeConfigOptions.Validate(nil); err != nil {
 		return err
 	}
-	if err := createNodeConfigOptions.CreateNodeFolder(); err != nil {
+	return createNodeConfigOptions.CreateNodeFolder()
+}
+
+func hasCSRCondition(conditions []certificates.CertificateSigningRequestCondition, t certificates.RequestConditionType) *certificates.CertificateSigningRequestCondition {
+	for i := range conditions {
+		if conditions[i].Type == t {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func (o NodeOptions) loadBootstrapClientCertificate(nodeConfigDir string, c kclientset.Interface, timeout time.Duration) (kclientset.Interface, error) {
+	nodeConfigPath := filepath.Join(nodeConfigDir, "node.kubeconfig")
+
+	// if the node config exists, try to use it or fail
+	if _, err := os.Stat(nodeConfigPath); err == nil {
+		kubeClientConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{ExplicitPath: nodeConfigPath}, &clientcmd.ConfigOverrides{}).ClientConfig()
+		if err != nil {
+			return nil, err
+		}
+		return kclientset.NewForConfig(kubeClientConfig)
+	}
+
+	clientCertPath := filepath.Join(nodeConfigDir, "master-client.crt")
+	clientKeyPath := filepath.Join(nodeConfigDir, "master-client.key")
+
+	// create and sign a client cert
+	privateKey, err := utilcert.NewPrivateKey()
+	if err != nil {
+		return nil, err
+	}
+	privateKeyData := utilcert.EncodePrivateKeyPEM(privateKey)
+	csrData, err := utilcert.MakeCSR(privateKey, &pkix.Name{
+		Organization: []string{"system:nodes"},
+		CommonName:   fmt.Sprintf("system:node:%s", o.NodeArgs.NodeName),
+	}, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	signingRequest := &certificates.CertificateSigningRequest{
+		ObjectMeta: kapi.ObjectMeta{
+			GenerateName: fmt.Sprintf("node-bootstrapper-%s-client-", safeSecretName(o.NodeArgs.NodeName)),
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request: csrData,
+		},
+	}
+
+	csr, err := c.Certificates().CertificateSigningRequests().Create(signingRequest)
```

Comment:
It is reentrant, uses generateName.  However, I think I'm going to step back from that to prevent bad loops creating too many CSRs, and use a nodeName -> name mapping.  I already check whether the cert is valid or not, but an admin has to delete the CSR for the node to continue.